### DISCUSSION TITLE
`Span.discharge` usage and caveats

### DISCUSSION THREAD

**Discussion by catostrophe** [trace4cats/trace4cats#750] [General]
@ybasket how do you see the usage of `DischargeSpan` with `EntryPoint`, especially when it continues a distributed trace passed via trace headers?

**Comment by ybasket** [comment]
`DischargeSpan` is meant to facilitate resource creation, especially with `IOLocal` where we have to provide a span when the `IOLocal` instance is created (but it is also helpful in `Kleisli` setups). Means that you would create your `Trace[F]` early in application lifetime – let's assume a small HTTP service with database – and then build your resources like HTTP routes using it. Fits very well the tagless final style of having context bounds for capabilities.

That could be done without `DischargeSpan`, either by employing a root span for initial phase or by using `EmptySpan`/`NoopSpan`. The former has the problem that if you perform any one-off actions that need tracing (database migrations, retrieving config once via HTTP), the spans created by that are all children of the one root span that won't be finished until the service is shut down. Empty spans lead to these one-off actions being lost as they're not collected anywhere. `DischargeSpan` is meant to solve that by creating a new root span for every different one-off action. I just noticed that it currently does that with setting a parent span context (whose data will never be collected, leading to warnings in Jaeger), will send a PR to fix.

`EntryPoint` will always continue received traces or create a new root, so it doesn't (to my understanding) interact at all with `DischargeSpan` other that, in the underlying implementation, they will share the same "storage" for their spans (`IOLocal`/`Kleisli`), but produce spans are related to spans of the other.

I updated my demo repo to include an [example of two service instances](https://github.com/ybasket/trace4cats-iolocal-demo/tree/main/src/main/scala/demo/even) calling each other via HTTP, it demonstrates the approach I describe.
